src/services/document_service.py

```diff
@@ -253,28 +253,76 @@ def get_document(self, reference: str, depth: int = 0, resolve_links: bool = Fal
             e.message = f"Failed to get document referenced with '{reference}'"
             raise e
 
+    def _get_node_to_update(self, reference: str, node_entity: Union[dict, list]) -> Union[Node, ListNode]:
+        """
+        Updating a document is done by fetching the node. This functions returns a node specified by reference.
+        Note: if the node specified by reference does not exist, it will be created if and only if the attribute is
+        specified as optional in the blueprint.
+        """
+        data_source_id, _reference = split_data_source_and_reference(reference)
+        reference_parts = split_reference(_reference)
+        parent_reference = "".join(reference_parts[:-1])
+        parent_node: Node = self.get_document(f"dmss://{data_source_id}/{parent_reference}", depth=0)
+        parent_blueprint_attribute_names = [attribute.name for attribute in parent_node.blueprint.attributes]
+        attribute_to_update = reference_parts[-1].strip(".").strip("[]")
+        node = parent_node.get_by_ref_part([attribute_to_update])
+
+        attribute_to_update_does_not_exist_in_parent_document = node is None
+
+        if (
+            attribute_to_update_does_not_exist_in_parent_document
+            and attribute_to_update in parent_blueprint_attribute_names
+        ):
+            attribute = [
+                attribute for attribute in parent_node.blueprint.attributes if attribute.name == attribute_to_update
+            ][0]
+
+            # We only want to add the attribute if it does not already exist AND is optional.
+            if not attribute.is_optional:
+                raise ValidationException(f"Could not update node. attribute '{attribute.name}' is not optional.")
+            if attribute.dimensions.dimensions != [""]:
+                node = ListNode(
+                    key=attribute_to_update,
+                    attribute=attribute,
+                    blueprint_provider=self.get_blueprint,
+                    entity=node_entity,
+                )
+            else:
+                node = Node(
+                    key=attribute_to_update,
+                    entity=node_entity,
+                    attribute=attribute,
+                    blueprint_provider=self.get_blueprint,
+                )
+        node.parent = parent_node
```

Comment:
`node` can still be none if `attribute_to_update not in parent_blueprint_attribute_names`. You should probably add some error handling here.

If I'm not wrong, there are two valid cases here.
1. `node is not None`
2. `node is None and attribute_to_update in [attribute.name for attribute in parent_node.blueprint.attributes if attribute.is_optional]`

All other cases should give an error.